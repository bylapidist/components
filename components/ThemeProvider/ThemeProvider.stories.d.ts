import { Meta, StoryObj } from '@storybook/react';
import { BaseProps } from '../types';
import { default as ThemeProvider, ThemeProviderProps } from './index';
declare const _default: Meta<typeof ThemeProvider>;
export default _default;
export declare const Light: StoryObj<BaseProps & ThemeProviderProps>;
export declare const Dark: StoryObj<BaseProps & ThemeProviderProps>;
