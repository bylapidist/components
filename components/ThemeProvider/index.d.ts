/// <reference types="react" />
import { Themes } from '@lapidist/styles';
import type { BaseProps } from '../types';
export interface ThemeProviderProps {
    readonly theme?: Themes;
}
declare const ThemeProvider: {
    ({ className, as: Component, testId, theme, ...restProps }: BaseProps & ThemeProviderProps): JSX.Element;
    displayName: string;
};
export default ThemeProvider;
