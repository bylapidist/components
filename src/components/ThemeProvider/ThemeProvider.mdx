import { Meta, ArgsTable, Story, Source, Canvas } from '@storybook/blocks';

import * as ThemeProviderStories from './ThemeProvider.stories.tsx';

<Meta of={ThemeProviderStories} />

# ThemeProvider

> A wrapper for providing theming support to all children.

## Example

<Source dark />

## Controlling theme

The `theme` prop controls the theme used by all child components. There are two available themes, defined as `light` and `dark`.

### Light

> Use the `light` theme by default, and for users who prefer light themes.

<Canvas>
    <Story of={ThemeProviderStories.Light} />
</Canvas>

### Dark

> Use the `dark` theme for users who prefer dark mode.

<Canvas>
    <Story of={ThemeProviderStories.Dark} />
</Canvas>

## Props

<ArgsTable />
