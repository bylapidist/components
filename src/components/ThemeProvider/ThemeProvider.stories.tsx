import * as React from 'react';
import type { Meta, StoryObj, StoryFn } from '@storybook/react';
import { BaseProps } from '../types';
import Text from '../Text';
import Box from '../Box';
import Heading from '../Heading';
import ThemeProvider, { ThemeProviderProps } from './index';
import Button from '../Button';

export default {
    title: 'Utility/ThemeProvider',
    component: Text,
    argTypes: {
        theme: { control: 'select', options: ['light', 'dark'] }
    }
} as Meta;

const defaultArgs: BaseProps & ThemeProviderProps = {
    as: 'div',
    testId: 'ThemeProvider',
    theme: 'light'
};

const DefaultTemplate: StoryFn<BaseProps & ThemeProviderProps> = (args) => (
    <ThemeProvider
        style={{
            backgroundColor: 'var(--color-grey__lightest)'
        }}
        {...args}
    >
        <Box gutter="large">
            <Box elevation="large" gutter="large">
                <Heading>Heading</Heading>
                <Text>
                    Nunc porttitor lectus ex, eu pharetra elit placerat non.
                    Suspendisse nec ultrices augue, et varius velit. Vestibulum
                    ante ipsum primis in faucibus orci luctus et ultrices
                    posuere cubilia curae; Morbi sollicitudin sed turpis et
                    vulputate. Nam consequat porttitor scelerisque. Nulla
                    ultricies enim at eros accumsan interdum. Aenean egestas
                    enim mi, nec tincidunt libero gravida et.
                </Text>
                <Box gutterY="large">
                    <Button level="primary">Click Me</Button>
                </Box>
            </Box>
        </Box>
    </ThemeProvider>
);

export const Light: StoryObj<BaseProps & ThemeProviderProps> =
    DefaultTemplate.bind({});
Light.args = {
    ...defaultArgs
};

export const Dark: StoryObj<BaseProps & ThemeProviderProps> =
    DefaultTemplate.bind({});
Dark.args = {
    ...defaultArgs,
    theme: 'dark'
};
