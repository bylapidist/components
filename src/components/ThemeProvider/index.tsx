import * as React from 'react';
import cx from 'classnames';
import { Themes } from '@lapidist/styles';
import '@lapidist/styles/dist/css/light.css';
import '@lapidist/styles/dist/css/dark.css';
import type { BaseProps } from '../types';

export interface ThemeProviderProps {
    readonly theme?: Themes;
}

export const ThemeContext: React.Context<Themes> =
    React.createContext<Themes>('light');

export const useThemeContext = (): Themes => React.useContext(ThemeContext);

const ThemeProvider = ({
    className,
    as: Component = 'div',
    testId = 'ThemeProvider',
    theme,
    ...restProps
}: BaseProps & ThemeProviderProps) => {
    const defaultTheme: Themes = useThemeContext();
    const [themeContext, setThemeContext] = React.useState<Themes>(
        theme || defaultTheme
    );

    React.useEffect(() => {
        setThemeContext(theme || defaultTheme);
    }, [theme]);

    React.useEffect(() => {
        if (
            (typeof window.matchMedia === 'function' &&
                window.matchMedia('(prefers-color-scheme: dark)')?.matches) ||
            localStorage.getItem('prefersDarkMode') === 'true'
        ) {
            setThemeContext('dark');
        }
    }, []);

    return (
        <Component
            className={cx(`theme--${themeContext}`, className)}
            data-testid={testId}
            {...restProps}
        />
    );
};

ThemeProvider.displayName = 'ThemeProvider';
export default ThemeProvider;
