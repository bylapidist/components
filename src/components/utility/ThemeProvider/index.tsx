import React from 'react';
import cx from 'classnames';
import { Themes } from '@lapidist/styles';
import '@lapidist/styles/dist/css/light.css';
import '@lapidist/styles/dist/css/dark.css';
import type { BaseProps } from '../../types';

export interface ThemeProviderProps {
    readonly theme?: Themes;
}

const ThemeProvider = ({
    className,
    as: Component = 'div',
    testId = 'ThemeProvider',
    theme = 'light',
    ...restProps
}: BaseProps & ThemeProviderProps) => (
    <Component
        className={cx(`theme--${theme}`, className)}
        data-testid={testId}
        {...restProps}
    />
);

ThemeProvider.displayName = 'ThemeProvider';
export default ThemeProvider;
