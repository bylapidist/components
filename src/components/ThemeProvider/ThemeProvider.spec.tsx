import React from 'react';
import { describe, expect, test } from 'vitest';
import { render, screen } from '@testing-library/react';
import ThemeProvider from './index';

const setup = (Component: React.ReactElement) => render(Component);

const themes: ('light' | 'dark')[] = ['light', 'dark'];

describe('ThemeProvider', () => {
    test('it works with defaults', () => {
        const { container } = setup(<ThemeProvider>Hello world</ThemeProvider>);
        expect(container.firstChild).toMatchSnapshot();
    });

    test.each(themes)('it works with themes', (theme) => {
        const { container } = setup(
            <ThemeProvider theme={theme}>Hello world</ThemeProvider>
        );
        expect(container.firstChild).toMatchSnapshot();
    });

    test('it works with default testId', () => {
        setup(<ThemeProvider>Hello world</ThemeProvider>);
        expect(screen.getByTestId('ThemeProvider')).toBeTruthy();
    });

    test('it works with specified testId', () => {
        setup(<ThemeProvider testId={'TestId'}>Hello world</ThemeProvider>);
        expect(screen.getByTestId('TestId')).toBeTruthy();
    });

    test('it works with as', () => {
        setup(<ThemeProvider as="span">Hello world</ThemeProvider>);
        expect(screen.getByTestId('ThemeProvider').nodeName).toBe('SPAN');
    });

    test('it works with className', () => {
        setup(<ThemeProvider className="test">Hello world</ThemeProvider>);
        expect(
            screen.getByTestId('ThemeProvider').classList.contains('test')
        ).toBeTruthy();
    });
});
