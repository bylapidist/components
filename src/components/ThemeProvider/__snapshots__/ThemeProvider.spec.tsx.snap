// Vitest Snapshot v1

exports[`ThemeProvider > it works with defaults 1`] = `
<div
  class="theme--dark"
  data-testid="ThemeProvider"
>
  Hello world
</div>
`;

exports[`ThemeProvider > it works with themes 1`] = `
<div
  class="theme--light"
  data-testid="ThemeProvider"
>
  Hello world
</div>
`;

exports[`ThemeProvider > it works with themes 2`] = `
<div
  class="theme--dark"
  data-testid="ThemeProvider"
>
  Hello world
</div>
`;
